MapboxNavigation/NavigationMapView.swift

```diff
@@ -671,6 +676,188 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
+    public var newFractionTraveled: Double = 0.0
+    
+    private struct RoutePoints {
+        var nestedList: [[[CLLocationCoordinate2D]]]
+        var flatList: [CLLocationCoordinate2D]
+    }
+    
+    private struct RouteLineGranularDistances {
+        var distance: Double
+        var distanceArray: [RouteLineDistancesIndex]
+    }
+    
+    private struct RouteLineDistancesIndex {
+        var point: CLLocationCoordinate2D
+        var distanceRemaining: Double
+    }
+    
+    private func initPrimaryRoutePoints(route: Route) {
+        primaryRoutePoints = parseRoutePoints(route: route)
+        primaryRouteLineGranularDistances = calculateRouteGranularDistances(coordinates: primaryRoutePoints?.flatList ?? [])
+    }
+    
+    /**
+    Tranform the route data into nested arrays of legs -> steps -> coordinates.
+    The first and last point of adjacent steps overlap and are duplicated.
+    */
+    private func parseRoutePoints(route: Route) -> RoutePoints {
+        let nestedList = route.legs.map { (routeLeg: RouteLeg) -> [[CLLocationCoordinate2D]] in
+            return routeLeg.steps.map { (routeStep: RouteStep) -> [CLLocationCoordinate2D] in
+                if let routeShape = routeStep.shape {
+                    if !routeShape.coordinates.isEmpty {
+                        return routeShape.coordinates
+                    } else { return [] }
+                } else {
+                    return []
+                }
+            }
+        }
+        let flatList = nestedList.flatMap { $0.flatMap { $0.flatMap { $0 } } }
+        return RoutePoints(nestedList: nestedList, flatList: flatList)
+    }
+    
+    /**
+    Find and cache the index of the upcoming [RouteLineDistancesIndex].
+    */
+    public func updateUpcomingRoutePointIndex(routeProgress: RouteProgress) {
+        let currentLegProgress = routeProgress.currentLegProgress
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        if currentLegProgress != nil && currentStepProgress != nil && primaryRoutePoints != nil {
+            let completeRoutePoints = primaryRoutePoints!
+            /**
+            Find the count of remaining points in the current step.
+            */
+            var allRemainingPoints = getSlicedLinePointsCount(currentLegProgress: currentLegProgress, currentStepProgress: currentStepProgress)
+            
+            /**
+            Add to the count of remaining points all of the remaining points on the current leg, after the current step.
+            */
+            let currentLegSteps = completeRoutePoints.nestedList[routeProgress.legIndex]
+            if currentLegProgress.stepIndex < currentLegSteps.count {
+                let startIndex = currentLegProgress.stepIndex + 1
+                let endIndex = currentLegSteps.count - 1
+                allRemainingPoints += currentLegSteps.prefix(endIndex).suffix(from: startIndex).flatMap{ $0.flatMap{ $0 } }.count
```

Comment:
As per [docs](https://developer.apple.com/documentation/swift/sequence/2905332-flatmap) `flatMap` was deprecated. Let's replace it with `compactMap`.

Reply:
Thanks for the suggestion, I'll replace the `flatMap `